rust/lance-index/src/scalar/rtree.rs

```diff
@@ -0,0 +1,1224 @@
+            .boxed();
+        let new_data = RecordBatchStreamAdapter::new(schema.clone(), stream);
+
+        Ok((
+            Box::pin(new_data),
+            BboxStreamStats {
+                null_map: null_rowids,
+                total_bbox,
+                num_items: num_non_null_rows,
+            },
+        ))
+    }
+
+    pub async fn write_index(
+        sorted_data: SendableRecordBatchStream,
+        num_items: usize,
+        bbox: BoundingBox,
+        store: &dyn IndexStore,
+        page_size: u32,
+    ) -> Result<()> {
+        let mut page_idx: u64 = 0;
+        let mut writer = store
+            .new_index_file(RTREE_PAGES_NAME, RTREE_PAGE_SCHEMA.clone())
+            .await?;
+
+        let mut page_offsets = vec![];
+        let mut curr_offset = 0;
+
+        if num_items > 0 {
+            let mut current_level = Some((sorted_data, num_items));
+            while let Some((mut data, num_items)) = current_level.take() {
+                if num_items <= page_size as usize {
+                    while let Some(batch) = data.next().await {
+                        let batch = batch?;
+                        page_offsets.push(curr_offset);
+                        curr_offset += batch.num_rows();
+                        train_rtree_page(batch, page_idx, writer.as_mut()).await?;
+                        page_idx += 1;
+                    }
+                } else {
+                    let mut next_level = vec![];
+                    let mut paged_source = chunk_concat_stream(data, page_size as usize);
+                    while let Some(batch) = paged_source.next().await {
+                        let batch = batch?;
+                        page_offsets.push(curr_offset);
+                        curr_offset += batch.num_rows();
+                        let encoded_batch =
+                            train_rtree_page(batch, page_idx, writer.as_mut()).await?;
+                        page_idx += 1;
+                        next_level.push(encoded_batch);
+                    }
+                    if !next_level.is_empty() {
+                        let next_num_items = next_level.len();
+                        current_level = Some((
+                            encoded_batches_into_batch_stream(next_level, page_size),
+                            next_num_items,
+                        ));
+                    }
+                }
+            }
+        }
+
+        writer
+            .finish_with_metadata(
+                RTreeMetadata::new(page_size, page_idx, page_offsets, num_items, bbox).into_map(),
+            )
+            .await?;
+
+        Ok(())
+    }
+
+    pub async fn write_nulls(store: &dyn IndexStore, null_map: RowIdTreeMap) -> Result<()> {
+        let mut writer = store
+            .new_index_file(RTREE_NULLS_NAME, RTREE_NULLS_SCHEMA.clone())
+            .await?;
+        let mut bytes = Vec::new();
+        null_map.serialize_into(&mut bytes)?;
```

Comment:
The serialization format of `RowIdTreeMap` is not stable according to the documentation of [`RowIdTreeMap.serialize_into`](https://docs.rs/lance-core/0.38.2/lance_core/utils/mask/struct.RowIdTreeMap.html#method.serialize_into). Persisting it in a file may introduce problems when the serialization format is changed.